boto3/dynamodb/table.py

```diff
@@ -0,0 +1,124 @@
+        super(TableResource, self).__init__(*args, **kwargs)
+
+    def batch_writer(self):
+        """Create a batch writer object.
+
+        This method creates a context manager for writing
+        objects to Amazon DynamoDB in batch.
+
+        The batch writer will automatically handle buffering and sending items
+        in batches.  In addition, the batch writer will also automatically
+        handle any unprocessed items and resend them as needed.  All you need
+        to do is call ``put_item`` for any items you want to add, and
+        ``delete_item`` for any items you want to delete.
+
+        Example usage::
+
+            with table.batch_writer() as batch:
+                for _ in xrange(1000000):
+                    batch.put_item(Item={'HashKey': '...',
+                                         'Otherstuff': '...'})
+                # You can also delete_items in a batch.
+                batch.delete_item(Key={'HashKey': 'SomeHashKey'})
+
+        """
+        return BatchWriter(self.name, self.meta.client)
+
+
+class BatchWriter(object):
+    """Automatically handle batch writes to DynamoDB for a single table."""
+    def __init__(self, table_name, client, flush_amount=25):
+        """
+
+        :type table_name: str
+        :param table_name: The name of the table.  The class handles
+            batch writes to a single table.
+
+        :type client: ``botocore.client.Client``
+        :param client: A botocore client.  Note this client
+            **must** have the dynamodb customizations applied
+            to it for transforming AttributeValues into the
+            wire protocol.  What this means in practice is that
+            you need to use a client that comes from a DynamoDB
+            resource if you're going to instantiate this class
+            directly, i.e
+            ``boto3.resource('dynamodb').Table('foo').meta.client``.
+
+        :type flush_amount: int
+        :param flush_amount: The number of items to keep in
+            a local buffer before sending a batch_write_item
+            request to DynamoDB.
+
+
+        """
+        self._table_name = table_name
+        self._client = client
+        self._items_buffer = []
+        self._flush_amount = flush_amount
+
+    def put_item(self, Item):
+        self._items_buffer.append({'PutRequest': {'Item': Item}})
+        self._flush_if_needed()
+
+    def delete_item(self, Key):
+        self._items_buffer.append({'DeleteRequest': {'Key': Key}})
+        self._flush_if_needed()
+
+    def _flush_if_needed(self):
+        if len(self._items_buffer) >= self._flush_amount:
+            self._flush()
+
+    def _flush(self):
+        logger.debug("Batch writer flushing pending items.")
+        response = self._client.batch_write_item(
+            RequestItems={self._table_name: self._items_buffer})
+        unprocessed_items = response['UnprocessedItems']
+
+        if unprocessed_items and unprocessed_items[self._table_name]:
```

Comment:
Is it guaranteed that the table_name is in the unprocessed_items? I would be wary of accessing it directly. Probably would use a ``get()``.